tmplctlr/controller_test.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2017 the lostromos Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tmplctlr
+
+import (
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var (
+	basicResource = &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"name": "dory",
+			},
+			"spec": map[string]interface{}{
+				"Name": "Dory",
+				"From": "Finding Nemo",
+				"By":   "Disney",
+			},
+		},
+	}
+
+	basicCR = &CustomResource{Resource: basicResource}
+
+	templates = []templateFile{
+		// T0.tmpl is a plain template file that just invokes T1.
+		{"0_base.tmpl", `--- {{template "file1.tmpl" . }}`},
+		// T1.tmpl defines a template, T1 that invokes T2.
+		{"file1.tmpl", `name: {{ .GetField "metadata" "name"  }}-configmap`},
+	}
+)
+
+// templateFile defines the contents of a template to be stored in a file, for testing.
+type templateFile struct {
+	name     string
+	contents string
+}
+
+func createTestDir(files []templateFile) string {
+	dir, err := ioutil.TempDir("", "template")
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, file := range files {
+		f, err := os.Create(filepath.Join(dir, file.name))
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer f.Close()
+		_, err = io.WriteString(f, file.contents)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	return dir
+}
+
+func ExampleController_ResourceAdded() {
+	dir := createTestDir(templates)
```

Comment:
Does go have the concept of setup/tearDown functionality like python unit testing does? Seems like we could clean up some code and make updating easier if that is the case.

Reply:
I like having each of these call `createTestDir` because then each test always has it's own tmp directory. We defer the deletion of the files in it so sometimes when the next test starts those files haven't been removed yet. We could make it more inline but that would add some slowdown to the tests.

Reply:
Yeah, this is a good place to use [TestMain](https://golang.org/pkg/testing/#hdr-Main).